File: go/leetcode/google/trees_and_graphs/maxPathSum.go
package google

func maxAB(a, b int) int {
	if a > b {
		return a
	}
	return b
}

const min = -1 * (1 << 30)

var max int

func maxPathSum(root *TreeNode) int {
	max = min
	sumPath(root)
	return max
}

func sumPath(root *TreeNode) int {
	sumRoot := root.Val
	var sumLeft, sumRight int
	localMax := sumRoot
	if root.Right != nil {
		sumRight = sumPath(root.Right)
		localMax = maxAB(localMax, sumRoot+sumRight)
	}
	if root.Left != nil {
		sumLeft = sumPath(root.Left)
		localMax = maxAB(localMax, sumRoot+sumLeft)
	}

	max = maxAB(max, sumRoot+sumLeft+sumRight)
	max = maxAB(max, localMax)
	return localMax
}

File: algoexpert/src/binary_trees/max_path_sum.py
maxSum = -1 * 1<<30

def maxPathSum(tree)-> int:
   global maxSum
   maxSum =  -1 * 1<<30
   maxPathSumHelper(tree)
   return maxSum

def maxPathSumHelper(tree) -> int:
    global maxSum

    response = tree.value
    fullPath = tree.value
    if tree.left is not None:
        left = maxPathSumHelper(tree.left)
        fullPath += left
        response = max(response, left + tree.value)
        maxSum = max(maxSum, left, left + tree.value)

    if tree.right is not None:
        right = maxPathSumHelper(tree.right)
        fullPath += right
        response = max(response, right + tree.value)
        maxSum = max(maxSum, right, right + tree.value)

    maxSum = max(maxSum, fullPath)

    return response